### PR TITLE
py-uxarray: new package plus dependencies

### PR DESCRIPTION
**Pull request by climbfuji**
## Description

This PR adds a new Python package `py-uxarray` (https://github.com/UXARRAY/uxarray). This package depends on many other Python packages, several of them don't exist in spack yet and are added here. Others need updating.

I am going to reach out to the "holoviz" developers and check with them about reviewing this PR, since many of the dependencies of `py-uxarray` are their packages. Maybe they will agree to be package maintainers, too. Please check https://github.com/holoviz/holoviz/issues/422

**I have two questions/issues which I need help with.**
1. For the update to `var/spack/repos/builtin/packages/py-panel/package.py`, should I reorder the dependencies similar to other Python packages? They seem to be different enough between the two "known" versions that it almost makes sense to keep them separate.
2. **Update 11/14/2024** I was able to resolve the import warning for dask dataframe that I reported earlier. The reason was a missing variant of `py-pyarrow` that caused the import of `py-dask-expr` to fail silently, and then `py-dask` complained that the dataframe query planning library was disabled because dask-expr was not installed. This is now resolved.

At yesterday's spack user meeting, we discussed the missing capability of the concretizer to map circular dependencies at runtime. Here, `py-dask` depends on `py-dask-expr` at runtime, and vice versa. This PR does not resolve this, instead it is up to the user (or the downstream dependency, in this case `py-uxarray`) to resolve this by requiring both (and `py-uxarray` does that correctly).

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-dask/package.py [hunk 1]
```diff
@@ -99,6 +99,9 @@ class PyDask(PythonPackage):
     depends_on("py-partd@0.3.10:", type=("build", "run"), when="@:2021.3.0 +dataframe")
     # The dependency on py-fsspec is non-optional starting version 2021.3.1
     depends_on("py-fsspec@0.6.0:", type=("build", "run"), when="@:2021.3.0 +dataframe")
+    # Starting with version 2024.3.0, dataframe requires a separate package py-dask-expr
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-dask/package.py:104]: According to the latest `requires.txt` (below), doesn't this mean `+array` is required when `+dataframe`?

```
[dataframe]
dask[array]
pandas>=2.0
dask-expr<1.2,>=1.1
```

Comment (climbfuji) [reply var/spack/repos/builtin/packages/py-dask/package.py:104]: Indeed. I am going back in history to find out when this was dependency was added, and then I'll add a conflict to catch this. It's unfortunately not the source of the problem I mentioned in item 2 above, but I found something else that I am exploring now.

Comment (climbfuji) [reply var/spack/repos/builtin/packages/py-dask/package.py:104]: @tldahlgren I fixed this in my latest commit, and I also fixed the `py-dask` import warning that `py-dask-expr` is not installed. See updated PR description. I am still fixing the style errors, but once those are done I am happy with this PR. I cloned the uxarray code separately and ran more than half of the tests that come with it against the spack-provided `uxarray` package.

**Comment by climbfuji** [comment]
@spackbot fix style

**Review by tldahlgren — APPROVED** [review]

**Comment by tldahlgren** [comment]
@haralmha @skosukhin Do you want to comment on this PR?

**Comment by climbfuji** [comment]
A gentle reminder for the reviewers, it would be nice if we could merge (or amend) this PR. Thank you!

**Comment by tldahlgren** [comment]
Pod error during generation of a stack.  Trying again:

@spackbot run pipeline

**Comment by tldahlgren** [comment]
It's been 3 weeks since my initial review and I'm not seeing additional comments. 🤞

**Comment by climbfuji** [comment]
Thank you @tldahlgren!

**Pull request merged**